src/Disks/VolumeJBOD.cpp

```diff
@@ -72,31 +83,61 @@ VolumeJBOD::VolumeJBOD(const VolumeJBOD & volume_jbod,
 
 DiskPtr VolumeJBOD::getDisk(size_t /* index */) const
 {
-    size_t start_from = last_used.fetch_add(1u, std::memory_order_acq_rel);
-    size_t index = start_from % disks.size();
-    return disks[index];
+    switch (load_balancing)
+    {
+        case VolumeLoadBalancing::ROUND_ROBIN:
+        {
+            size_t start_from = last_used.fetch_add(1u, std::memory_order_acq_rel);
+            size_t index = start_from % disks.size();
+            return disks[index];
+        }
+        case VolumeLoadBalancing::LEAST_USED:
+        {
+            std::lock_guard lock(mutex);
+            return disks_by_size.top().disk;
+        }
+    }
+    __builtin_unreachable();
 }
 
 ReservationPtr VolumeJBOD::reserve(UInt64 bytes)
 {
     /// This volume can not store data which size is greater than `max_data_part_size`
     /// to ensure that parts of size greater than that go to another volume(s).
-
     if (max_data_part_size != 0 && bytes > max_data_part_size)
         return {};
 
-    size_t start_from = last_used.fetch_add(1u, std::memory_order_acq_rel);
-    size_t disks_num = disks.size();
-    for (size_t i = 0; i < disks_num; ++i)
+    switch (load_balancing)
     {
-        size_t index = (start_from + i) % disks_num;
+        case VolumeLoadBalancing::ROUND_ROBIN:
+        {
+            size_t start_from = last_used.fetch_add(1u, std::memory_order_acq_rel);
+            size_t disks_num = disks.size();
+            for (size_t i = 0; i < disks_num; ++i)
+            {
+                size_t index = (start_from + i) % disks_num;
+
+                auto reservation = disks[index]->reserve(bytes);
+
+                if (reservation)
+                    return reservation;
+            }
+            return {};
+        }
+        case VolumeLoadBalancing::LEAST_USED:
+        {
+            std::lock_guard lock(mutex);
 
-        auto reservation = disks[index]->reserve(bytes);
+            DiskWithSize disk = disks_by_size.top();
+            ReservationPtr reservation = disk.reserve(bytes);
```

Comment:
This is still wrong. Now pop() will get a heap with violated heap property. Right order: pop(), modify key, push() it back.

Reply:
Nice catch, will be fixed in the next rebase.